diff --git a/avm-transpiler/src/transpile.rs b/avm-transpiler/src/transpile.rs
index 9c9c356ce4..dfe08a684c 100644
--- a/avm-transpiler/src/transpile.rs
+++ b/avm-transpiler/src/transpile.rs
@@ -440,6 +440,7 @@ pub fn brillig_to_avm(brillig_bytecode: &[BrilligOpcode<FieldElement>]) -> (Vec<
                     &mut unresolved_jumps,
                 );
             }
+            &acvm::acir::brillig::Opcode::JumpIfNot { .. } => todo!(),
         }

         // Increment the AVM program counter.
@@ -1194,7 +1195,7 @@ fn handle_black_box_function(
                 ..Default::default()
             });
         }
-        BlackBoxOp::Poseidon2Permutation { message, output } => {
+        BlackBoxOp::Poseidon2Permutation { message, output, .. } => {
             // We'd love to validate the input size, but it's not known at compile time.
             assert_eq!(output.size, 4, "Poseidon2Permutation output size must be 4!");
             let input_state_offset = message.pointer.to_usize();
