From dd3679d25ce572ccadfa46b5a6028315e2de3370 Mon Sep 17 00:00:00 2001
From: TomAFrench <tom@tomfren.ch>
Date: Sat, 15 Mar 2025 15:36:12 +0000
Subject: [PATCH 1/4] chore: turn on `skipLibCheck`

---
 tooling/noir_codegen/tsconfig.json | 1 +
 1 file changed, 1 insertion(+)

diff --git a/tooling/noir_codegen/tsconfig.json b/tooling/noir_codegen/tsconfig.json
index 30dd2a7..a2712fd 100644
--- a/tooling/noir_codegen/tsconfig.json
+++ b/tooling/noir_codegen/tsconfig.json
@@ -10,6 +10,7 @@
     "resolveJsonModule": true,
     "strict": true,
     "noImplicitAny": false,
+    "skipLibCheck": true
   },
   "include": [
     "src/**/*.ts"
-- 
2.43.0

From 7cdbc4d8779c16e2acb704abfbbda84e66a12986 Mon Sep 17 00:00:00 2001
From: dbanks12 <david@aztec-labs.com>
Date: Thu, 21 Aug 2025 17:02:22 +0000
Subject: [PATCH 2/4] feat!: force inliner aggressiveness to 0 for for
 public_dispatch contract fn

---
 compiler/noirc_driver/src/lib.rs | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/compiler/noirc_driver/src/lib.rs b/compiler/noirc_driver/src/lib.rs
index e4da2da..ff2cdaa 100644
--- a/compiler/noirc_driver/src/lib.rs
+++ b/compiler/noirc_driver/src/lib.rs
@@ -645,6 +645,10 @@ fn compile_contract_inner(
             }
         };
 
+        if name == "public_dispatch" {
+            options.inliner_aggressiveness = 0;
+        }
+
         let function = match compile_no_check(context, &options, function_id, None, true) {
             Ok(function) => function,
             Err(new_error) => {
-- 
2.43.0

From a4a0a9098327e412de760a857339cd6bfe04bac3 Mon Sep 17 00:00:00 2001
From: Tom French <15848336+TomAFrench@users.noreply.github.com>
Date: Fri, 5 Sep 2025 13:07:32 +0000
Subject: [PATCH 4/4] enable mem2reg pass

---
 compiler/noirc_evaluator/src/ssa/mod.rs | 4 +---
 1 file changed, 1 insertion(+), 3 deletions(-)

diff --git a/compiler/noirc_evaluator/src/ssa/mod.rs b/compiler/noirc_evaluator/src/ssa/mod.rs
index 7edcb14..b8bdc27 100644
--- a/compiler/noirc_evaluator/src/ssa/mod.rs
+++ b/compiler/noirc_evaluator/src/ssa/mod.rs
@@ -109,9 +109,7 @@ pub fn primary_passes(options: &SsaEvaluatorOptions) -> Vec<SsaPass<'_>> {
         SsaPass::new(Ssa::defunctionalize, "Defunctionalization"),
         SsaPass::new(Ssa::inline_simple_functions, "Inlining simple functions")
             .and_then(Ssa::remove_unreachable_functions),
-        // BUG: Enabling this mem2reg causes an integration test failure in aztec-package; see:
-        // https://github.com/AztecProtocol/aztec-packages/pull/11294#issuecomment-2622809518
-        //SsaPass::new(Ssa::mem2reg, "Mem2Reg (1st)"),
+        SsaPass::new(Ssa::mem2reg, "Mem2Reg"),
         SsaPass::new(Ssa::remove_paired_rc, "Removing Paired rc_inc & rc_decs"),
         SsaPass::new_try(
             move |ssa| ssa.preprocess_functions(options.inliner_aggressiveness),
-- 
2.43.0

