diff --git a/lib/packet/encode.js b/lib/packet/encode.js
index f1eb92f606a9df570975ba1ce9c4b599bfc2c432..1d3555b04c1fa07465927f3b9c1a3a9a124e4762 100644
--- a/lib/packet/encode.js
+++ b/lib/packet/encode.js
@@ -70,9 +70,10 @@ export function decodeHeader(srcId, maskingIv, data) {
 }
 // authdata
 export function encodeWhoAreYouAuthdata(authdata) {
-    // Pad sequence to 8 bytes
+    // Pad sequence to 8 bytes (big-endian, right-aligned)
     const seqBytes = new Uint8Array(8);
-    seqBytes.set(bigintToBytes(authdata.enrSeq), 0);
+    const bytes = bigintToBytes(authdata.enrSeq);
+    seqBytes.set(bytes, 8 - bytes.length);
     return concatBytes(authdata.idNonce, seqBytes);
 }
 export function encodeMessageAuthdata(authdata) {
