diff --git a/_esm/utils/transaction/parseTransaction.js b/_esm/utils/transaction/parseTransaction.js
index 6390fb1bf43b2d93af2836443d669e4ecb07d11e..47cb6448869138da57919338dd51dfeae6b923be 100644
--- a/_esm/utils/transaction/parseTransaction.js
+++ b/_esm/utils/transaction/parseTransaction.js
@@ -1,350 +1,375 @@
-import { InvalidAddressError, } from '../../errors/address.js';
-import { InvalidLegacyVError, InvalidSerializedTransactionError, } from '../../errors/transaction.js';
+import { InvalidAddressError } from '../../errors/address.js';
+import {
+  InvalidLegacyVError,
+  InvalidSerializedTransactionError,
+} from '../../errors/transaction.js';
 import { isAddress } from '../address/isAddress.js';
 import { toBlobSidecars } from '../blob/toBlobSidecars.js';
 import { isHex } from '../data/isHex.js';
 import { padHex } from '../data/pad.js';
 import { trim } from '../data/trim.js';
-import { hexToBigInt, hexToNumber, } from '../encoding/fromHex.js';
+import { hexToBigInt, hexToNumber } from '../encoding/fromHex.js';
 import { fromRlp } from '../encoding/fromRlp.js';
 import { isHash } from '../hash/isHash.js';
-import { assertTransactionEIP1559, assertTransactionEIP2930, assertTransactionEIP4844, assertTransactionEIP7702, assertTransactionLegacy, } from './assertTransaction.js';
-import { getSerializedTransactionType, } from './getSerializedTransactionType.js';
+import {
+  assertTransactionEIP1559,
+  assertTransactionEIP2930,
+  assertTransactionEIP4844,
+  assertTransactionEIP7702,
+  assertTransactionLegacy,
+} from './assertTransaction.js';
+import { getSerializedTransactionType } from './getSerializedTransactionType.js';
 export function parseTransaction(serializedTransaction) {
-    const type = getSerializedTransactionType(serializedTransaction);
-    if (type === 'eip1559')
-        return parseTransactionEIP1559(serializedTransaction);
-    if (type === 'eip2930')
-        return parseTransactionEIP2930(serializedTransaction);
-    if (type === 'eip4844')
-        return parseTransactionEIP4844(serializedTransaction);
-    if (type === 'eip7702')
-        return parseTransactionEIP7702(serializedTransaction);
-    return parseTransactionLegacy(serializedTransaction);
+  const type = getSerializedTransactionType(serializedTransaction);
+  if (type === 'eip1559') return parseTransactionEIP1559(serializedTransaction);
+  if (type === 'eip2930') return parseTransactionEIP2930(serializedTransaction);
+  if (type === 'eip4844') return parseTransactionEIP4844(serializedTransaction);
+  if (type === 'eip7702') return parseTransactionEIP7702(serializedTransaction);
+  return parseTransactionLegacy(serializedTransaction);
 }
 function parseTransactionEIP7702(serializedTransaction) {
-    const transactionArray = toTransactionArray(serializedTransaction);
-    const [chainId, nonce, maxPriorityFeePerGas, maxFeePerGas, gas, to, value, data, accessList, authorizationList, v, r, s,] = transactionArray;
-    if (transactionArray.length !== 10 && transactionArray.length !== 13)
-        throw new InvalidSerializedTransactionError({
-            attributes: {
-                chainId,
-                nonce,
-                maxPriorityFeePerGas,
-                maxFeePerGas,
-                gas,
-                to,
-                value,
-                data,
-                accessList,
-                authorizationList,
-                ...(transactionArray.length > 9
-                    ? {
-                        v,
-                        r,
-                        s,
-                    }
-                    : {}),
-            },
-            serializedTransaction,
-            type: 'eip7702',
-        });
-    const transaction = {
-        chainId: hexToNumber(chainId),
-        type: 'eip7702',
-    };
-    if (isHex(to) && to !== '0x')
-        transaction.to = to;
-    if (isHex(gas) && gas !== '0x')
-        transaction.gas = hexToBigInt(gas);
-    if (isHex(data) && data !== '0x')
-        transaction.data = data;
-    if (isHex(nonce) && nonce !== '0x')
-        transaction.nonce = hexToNumber(nonce);
-    if (isHex(value) && value !== '0x')
-        transaction.value = hexToBigInt(value);
-    if (isHex(maxFeePerGas) && maxFeePerGas !== '0x')
-        transaction.maxFeePerGas = hexToBigInt(maxFeePerGas);
-    if (isHex(maxPriorityFeePerGas) && maxPriorityFeePerGas !== '0x')
-        transaction.maxPriorityFeePerGas = hexToBigInt(maxPriorityFeePerGas);
-    if (accessList.length !== 0 && accessList !== '0x')
-        transaction.accessList = parseAccessList(accessList);
-    if (authorizationList.length !== 0 && authorizationList !== '0x')
-        transaction.authorizationList = parseAuthorizationList(authorizationList);
-    assertTransactionEIP7702(transaction);
-    const signature = transactionArray.length === 13
-        ? parseEIP155Signature(transactionArray)
-        : undefined;
-    return { ...signature, ...transaction };
+  const transactionArray = toTransactionArray(serializedTransaction);
+  const [
+    chainId,
+    nonce,
+    maxPriorityFeePerGas,
+    maxFeePerGas,
+    gas,
+    to,
+    value,
+    data,
+    accessList,
+    authorizationList,
+    v,
+    r,
+    s,
+  ] = transactionArray;
+  if (transactionArray.length !== 10 && transactionArray.length !== 13)
+    throw new InvalidSerializedTransactionError({
+      attributes: {
+        chainId,
+        nonce,
+        maxPriorityFeePerGas,
+        maxFeePerGas,
+        gas,
+        to,
+        value,
+        data,
+        accessList,
+        authorizationList,
+        ...(transactionArray.length > 9
+          ? {
+              v,
+              r,
+              s,
+            }
+          : {}),
+      },
+      serializedTransaction,
+      type: 'eip7702',
+    });
+  const transaction = {
+    chainId: hexToNumber(chainId),
+    type: 'eip7702',
+  };
+  if (isHex(to) && to !== '0x') transaction.to = to;
+  if (isHex(gas) && gas !== '0x') transaction.gas = hexToBigInt(gas);
+  if (isHex(data) && data !== '0x') transaction.data = data;
+  if (isHex(nonce) && nonce !== '0x') transaction.nonce = hexToNumber(nonce);
+  if (isHex(value) && value !== '0x') transaction.value = hexToBigInt(value);
+  if (isHex(maxFeePerGas) && maxFeePerGas !== '0x')
+    transaction.maxFeePerGas = hexToBigInt(maxFeePerGas);
+  if (isHex(maxPriorityFeePerGas) && maxPriorityFeePerGas !== '0x')
+    transaction.maxPriorityFeePerGas = hexToBigInt(maxPriorityFeePerGas);
+  if (accessList.length !== 0 && accessList !== '0x')
+    transaction.accessList = parseAccessList(accessList);
+  if (authorizationList.length !== 0 && authorizationList !== '0x')
+    transaction.authorizationList = parseAuthorizationList(authorizationList);
+  assertTransactionEIP7702(transaction);
+  const signature =
+    transactionArray.length === 13
+      ? parseEIP155Signature(transactionArray)
+      : undefined;
+  return { ...signature, ...transaction };
 }
 function parseTransactionEIP4844(serializedTransaction) {
-    const transactionOrWrapperArray = toTransactionArray(serializedTransaction);
-    const hasNetworkWrapper = transactionOrWrapperArray.length === 4;
-    const transactionArray = hasNetworkWrapper
-        ? transactionOrWrapperArray[0]
-        : transactionOrWrapperArray;
-    const wrapperArray = hasNetworkWrapper
-        ? transactionOrWrapperArray.slice(1)
-        : [];
-    const [chainId, nonce, maxPriorityFeePerGas, maxFeePerGas, gas, to, value, data, accessList, maxFeePerBlobGas, blobVersionedHashes, v, r, s,] = transactionArray;
-    const [blobs, commitments, proofs] = wrapperArray;
-    if (!(transactionArray.length === 11 || transactionArray.length === 14))
-        throw new InvalidSerializedTransactionError({
-            attributes: {
-                chainId,
-                nonce,
-                maxPriorityFeePerGas,
-                maxFeePerGas,
-                gas,
-                to,
-                value,
-                data,
-                accessList,
-                ...(transactionArray.length > 9
-                    ? {
-                        v,
-                        r,
-                        s,
-                    }
-                    : {}),
-            },
-            serializedTransaction,
-            type: 'eip4844',
-        });
-    const transaction = {
-        blobVersionedHashes: blobVersionedHashes,
-        chainId: hexToNumber(chainId),
-        type: 'eip4844',
-    };
-    if (isHex(to) && to !== '0x')
-        transaction.to = to;
-    if (isHex(gas) && gas !== '0x')
-        transaction.gas = hexToBigInt(gas);
-    if (isHex(data) && data !== '0x')
-        transaction.data = data;
-    if (isHex(nonce) && nonce !== '0x')
-        transaction.nonce = hexToNumber(nonce);
-    if (isHex(value) && value !== '0x')
-        transaction.value = hexToBigInt(value);
-    if (isHex(maxFeePerBlobGas) && maxFeePerBlobGas !== '0x')
-        transaction.maxFeePerBlobGas = hexToBigInt(maxFeePerBlobGas);
-    if (isHex(maxFeePerGas) && maxFeePerGas !== '0x')
-        transaction.maxFeePerGas = hexToBigInt(maxFeePerGas);
-    if (isHex(maxPriorityFeePerGas) && maxPriorityFeePerGas !== '0x')
-        transaction.maxPriorityFeePerGas = hexToBigInt(maxPriorityFeePerGas);
-    if (accessList.length !== 0 && accessList !== '0x')
-        transaction.accessList = parseAccessList(accessList);
-    if (blobs && commitments && proofs)
-        transaction.sidecars = toBlobSidecars({
-            blobs: blobs,
-            commitments: commitments,
-            proofs: proofs,
-        });
-    assertTransactionEIP4844(transaction);
-    const signature = transactionArray.length === 14
-        ? parseEIP155Signature(transactionArray)
-        : undefined;
-    return { ...signature, ...transaction };
+  const transactionOrWrapperArray = toTransactionArray(serializedTransaction);
+  const hasNetworkWrapper = transactionOrWrapperArray.length === 5;
+  const transactionArray = hasNetworkWrapper
+    ? transactionOrWrapperArray[0]
+    : transactionOrWrapperArray;
+  const wrapperArray = hasNetworkWrapper
+    ? transactionOrWrapperArray.slice(1)
+    : [];
+  const [
+    chainId,
+    nonce,
+    maxPriorityFeePerGas,
+    maxFeePerGas,
+    gas,
+    to,
+    value,
+    data,
+    accessList,
+    maxFeePerBlobGas,
+    blobVersionedHashes,
+    v,
+    r,
+    s,
+  ] = transactionArray;
+  const [blobs, commitments, proofs] = wrapperArray;
+  if (!(transactionArray.length === 11 || transactionArray.length === 14))
+    throw new InvalidSerializedTransactionError({
+      attributes: {
+        chainId,
+        nonce,
+        maxPriorityFeePerGas,
+        maxFeePerGas,
+        gas,
+        to,
+        value,
+        data,
+        accessList,
+        ...(transactionArray.length > 9
+          ? {
+              v,
+              r,
+              s,
+            }
+          : {}),
+      },
+      serializedTransaction,
+      type: 'eip4844',
+    });
+  const transaction = {
+    blobVersionedHashes: blobVersionedHashes,
+    chainId: hexToNumber(chainId),
+    type: 'eip4844',
+  };
+  if (isHex(to) && to !== '0x') transaction.to = to;
+  if (isHex(gas) && gas !== '0x') transaction.gas = hexToBigInt(gas);
+  if (isHex(data) && data !== '0x') transaction.data = data;
+  if (isHex(nonce) && nonce !== '0x') transaction.nonce = hexToNumber(nonce);
+  if (isHex(value) && value !== '0x') transaction.value = hexToBigInt(value);
+  if (isHex(maxFeePerBlobGas) && maxFeePerBlobGas !== '0x')
+    transaction.maxFeePerBlobGas = hexToBigInt(maxFeePerBlobGas);
+  if (isHex(maxFeePerGas) && maxFeePerGas !== '0x')
+    transaction.maxFeePerGas = hexToBigInt(maxFeePerGas);
+  if (isHex(maxPriorityFeePerGas) && maxPriorityFeePerGas !== '0x')
+    transaction.maxPriorityFeePerGas = hexToBigInt(maxPriorityFeePerGas);
+  if (accessList.length !== 0 && accessList !== '0x')
+    transaction.accessList = parseAccessList(accessList);
+  if (blobs && commitments && proofs)
+    transaction.sidecars = toBlobSidecars({
+      blobs: blobs,
+      commitments: commitments,
+      proofs: proofs,
+    });
+  assertTransactionEIP4844(transaction);
+  const signature =
+    transactionArray.length === 14
+      ? parseEIP155Signature(transactionArray)
+      : undefined;
+  return { ...signature, ...transaction };
 }
 function parseTransactionEIP1559(serializedTransaction) {
-    const transactionArray = toTransactionArray(serializedTransaction);
-    const [chainId, nonce, maxPriorityFeePerGas, maxFeePerGas, gas, to, value, data, accessList, v, r, s,] = transactionArray;
-    if (!(transactionArray.length === 9 || transactionArray.length === 12))
-        throw new InvalidSerializedTransactionError({
-            attributes: {
-                chainId,
-                nonce,
-                maxPriorityFeePerGas,
-                maxFeePerGas,
-                gas,
-                to,
-                value,
-                data,
-                accessList,
-                ...(transactionArray.length > 9
-                    ? {
-                        v,
-                        r,
-                        s,
-                    }
-                    : {}),
-            },
-            serializedTransaction,
-            type: 'eip1559',
-        });
-    const transaction = {
-        chainId: hexToNumber(chainId),
-        type: 'eip1559',
-    };
-    if (isHex(to) && to !== '0x')
-        transaction.to = to;
-    if (isHex(gas) && gas !== '0x')
-        transaction.gas = hexToBigInt(gas);
-    if (isHex(data) && data !== '0x')
-        transaction.data = data;
-    if (isHex(nonce) && nonce !== '0x')
-        transaction.nonce = hexToNumber(nonce);
-    if (isHex(value) && value !== '0x')
-        transaction.value = hexToBigInt(value);
-    if (isHex(maxFeePerGas) && maxFeePerGas !== '0x')
-        transaction.maxFeePerGas = hexToBigInt(maxFeePerGas);
-    if (isHex(maxPriorityFeePerGas) && maxPriorityFeePerGas !== '0x')
-        transaction.maxPriorityFeePerGas = hexToBigInt(maxPriorityFeePerGas);
-    if (accessList.length !== 0 && accessList !== '0x')
-        transaction.accessList = parseAccessList(accessList);
-    assertTransactionEIP1559(transaction);
-    const signature = transactionArray.length === 12
-        ? parseEIP155Signature(transactionArray)
-        : undefined;
-    return { ...signature, ...transaction };
+  const transactionArray = toTransactionArray(serializedTransaction);
+  const [
+    chainId,
+    nonce,
+    maxPriorityFeePerGas,
+    maxFeePerGas,
+    gas,
+    to,
+    value,
+    data,
+    accessList,
+    v,
+    r,
+    s,
+  ] = transactionArray;
+  if (!(transactionArray.length === 9 || transactionArray.length === 12))
+    throw new InvalidSerializedTransactionError({
+      attributes: {
+        chainId,
+        nonce,
+        maxPriorityFeePerGas,
+        maxFeePerGas,
+        gas,
+        to,
+        value,
+        data,
+        accessList,
+        ...(transactionArray.length > 9
+          ? {
+              v,
+              r,
+              s,
+            }
+          : {}),
+      },
+      serializedTransaction,
+      type: 'eip1559',
+    });
+  const transaction = {
+    chainId: hexToNumber(chainId),
+    type: 'eip1559',
+  };
+  if (isHex(to) && to !== '0x') transaction.to = to;
+  if (isHex(gas) && gas !== '0x') transaction.gas = hexToBigInt(gas);
+  if (isHex(data) && data !== '0x') transaction.data = data;
+  if (isHex(nonce) && nonce !== '0x') transaction.nonce = hexToNumber(nonce);
+  if (isHex(value) && value !== '0x') transaction.value = hexToBigInt(value);
+  if (isHex(maxFeePerGas) && maxFeePerGas !== '0x')
+    transaction.maxFeePerGas = hexToBigInt(maxFeePerGas);
+  if (isHex(maxPriorityFeePerGas) && maxPriorityFeePerGas !== '0x')
+    transaction.maxPriorityFeePerGas = hexToBigInt(maxPriorityFeePerGas);
+  if (accessList.length !== 0 && accessList !== '0x')
+    transaction.accessList = parseAccessList(accessList);
+  assertTransactionEIP1559(transaction);
+  const signature =
+    transactionArray.length === 12
+      ? parseEIP155Signature(transactionArray)
+      : undefined;
+  return { ...signature, ...transaction };
 }
 function parseTransactionEIP2930(serializedTransaction) {
-    const transactionArray = toTransactionArray(serializedTransaction);
-    const [chainId, nonce, gasPrice, gas, to, value, data, accessList, v, r, s] = transactionArray;
-    if (!(transactionArray.length === 8 || transactionArray.length === 11))
-        throw new InvalidSerializedTransactionError({
-            attributes: {
-                chainId,
-                nonce,
-                gasPrice,
-                gas,
-                to,
-                value,
-                data,
-                accessList,
-                ...(transactionArray.length > 8
-                    ? {
-                        v,
-                        r,
-                        s,
-                    }
-                    : {}),
-            },
-            serializedTransaction,
-            type: 'eip2930',
-        });
-    const transaction = {
-        chainId: hexToNumber(chainId),
-        type: 'eip2930',
-    };
-    if (isHex(to) && to !== '0x')
-        transaction.to = to;
-    if (isHex(gas) && gas !== '0x')
-        transaction.gas = hexToBigInt(gas);
-    if (isHex(data) && data !== '0x')
-        transaction.data = data;
-    if (isHex(nonce) && nonce !== '0x')
-        transaction.nonce = hexToNumber(nonce);
-    if (isHex(value) && value !== '0x')
-        transaction.value = hexToBigInt(value);
-    if (isHex(gasPrice) && gasPrice !== '0x')
-        transaction.gasPrice = hexToBigInt(gasPrice);
-    if (accessList.length !== 0 && accessList !== '0x')
-        transaction.accessList = parseAccessList(accessList);
-    assertTransactionEIP2930(transaction);
-    const signature = transactionArray.length === 11
-        ? parseEIP155Signature(transactionArray)
-        : undefined;
-    return { ...signature, ...transaction };
+  const transactionArray = toTransactionArray(serializedTransaction);
+  const [chainId, nonce, gasPrice, gas, to, value, data, accessList, v, r, s] =
+    transactionArray;
+  if (!(transactionArray.length === 8 || transactionArray.length === 11))
+    throw new InvalidSerializedTransactionError({
+      attributes: {
+        chainId,
+        nonce,
+        gasPrice,
+        gas,
+        to,
+        value,
+        data,
+        accessList,
+        ...(transactionArray.length > 8
+          ? {
+              v,
+              r,
+              s,
+            }
+          : {}),
+      },
+      serializedTransaction,
+      type: 'eip2930',
+    });
+  const transaction = {
+    chainId: hexToNumber(chainId),
+    type: 'eip2930',
+  };
+  if (isHex(to) && to !== '0x') transaction.to = to;
+  if (isHex(gas) && gas !== '0x') transaction.gas = hexToBigInt(gas);
+  if (isHex(data) && data !== '0x') transaction.data = data;
+  if (isHex(nonce) && nonce !== '0x') transaction.nonce = hexToNumber(nonce);
+  if (isHex(value) && value !== '0x') transaction.value = hexToBigInt(value);
+  if (isHex(gasPrice) && gasPrice !== '0x')
+    transaction.gasPrice = hexToBigInt(gasPrice);
+  if (accessList.length !== 0 && accessList !== '0x')
+    transaction.accessList = parseAccessList(accessList);
+  assertTransactionEIP2930(transaction);
+  const signature =
+    transactionArray.length === 11
+      ? parseEIP155Signature(transactionArray)
+      : undefined;
+  return { ...signature, ...transaction };
 }
 function parseTransactionLegacy(serializedTransaction) {
-    const transactionArray = fromRlp(serializedTransaction, 'hex');
-    const [nonce, gasPrice, gas, to, value, data, chainIdOrV_, r, s] = transactionArray;
-    if (!(transactionArray.length === 6 || transactionArray.length === 9))
-        throw new InvalidSerializedTransactionError({
-            attributes: {
-                nonce,
-                gasPrice,
-                gas,
-                to,
-                value,
-                data,
-                ...(transactionArray.length > 6
-                    ? {
-                        v: chainIdOrV_,
-                        r,
-                        s,
-                    }
-                    : {}),
-            },
-            serializedTransaction,
-            type: 'legacy',
-        });
-    const transaction = {
-        type: 'legacy',
-    };
-    if (isHex(to) && to !== '0x')
-        transaction.to = to;
-    if (isHex(gas) && gas !== '0x')
-        transaction.gas = hexToBigInt(gas);
-    if (isHex(data) && data !== '0x')
-        transaction.data = data;
-    if (isHex(nonce) && nonce !== '0x')
-        transaction.nonce = hexToNumber(nonce);
-    if (isHex(value) && value !== '0x')
-        transaction.value = hexToBigInt(value);
-    if (isHex(gasPrice) && gasPrice !== '0x')
-        transaction.gasPrice = hexToBigInt(gasPrice);
-    assertTransactionLegacy(transaction);
-    if (transactionArray.length === 6)
-        return transaction;
-    const chainIdOrV = isHex(chainIdOrV_) && chainIdOrV_ !== '0x'
-        ? hexToBigInt(chainIdOrV_)
-        : 0n;
-    if (s === '0x' && r === '0x') {
-        if (chainIdOrV > 0)
-            transaction.chainId = Number(chainIdOrV);
-        return transaction;
-    }
-    const v = chainIdOrV;
-    const chainId = Number((v - 35n) / 2n);
-    if (chainId > 0)
-        transaction.chainId = chainId;
-    else if (v !== 27n && v !== 28n)
-        throw new InvalidLegacyVError({ v });
-    transaction.v = v;
-    transaction.s = s;
-    transaction.r = r;
-    transaction.yParity = v % 2n === 0n ? 1 : 0;
+  const transactionArray = fromRlp(serializedTransaction, 'hex');
+  const [nonce, gasPrice, gas, to, value, data, chainIdOrV_, r, s] =
+    transactionArray;
+  if (!(transactionArray.length === 6 || transactionArray.length === 9))
+    throw new InvalidSerializedTransactionError({
+      attributes: {
+        nonce,
+        gasPrice,
+        gas,
+        to,
+        value,
+        data,
+        ...(transactionArray.length > 6
+          ? {
+              v: chainIdOrV_,
+              r,
+              s,
+            }
+          : {}),
+      },
+      serializedTransaction,
+      type: 'legacy',
+    });
+  const transaction = {
+    type: 'legacy',
+  };
+  if (isHex(to) && to !== '0x') transaction.to = to;
+  if (isHex(gas) && gas !== '0x') transaction.gas = hexToBigInt(gas);
+  if (isHex(data) && data !== '0x') transaction.data = data;
+  if (isHex(nonce) && nonce !== '0x') transaction.nonce = hexToNumber(nonce);
+  if (isHex(value) && value !== '0x') transaction.value = hexToBigInt(value);
+  if (isHex(gasPrice) && gasPrice !== '0x')
+    transaction.gasPrice = hexToBigInt(gasPrice);
+  assertTransactionLegacy(transaction);
+  if (transactionArray.length === 6) return transaction;
+  const chainIdOrV =
+    isHex(chainIdOrV_) && chainIdOrV_ !== '0x' ? hexToBigInt(chainIdOrV_) : 0n;
+  if (s === '0x' && r === '0x') {
+    if (chainIdOrV > 0) transaction.chainId = Number(chainIdOrV);
     return transaction;
+  }
+  const v = chainIdOrV;
+  const chainId = Number((v - 35n) / 2n);
+  if (chainId > 0) transaction.chainId = chainId;
+  else if (v !== 27n && v !== 28n) throw new InvalidLegacyVError({ v });
+  transaction.v = v;
+  transaction.s = s;
+  transaction.r = r;
+  transaction.yParity = v % 2n === 0n ? 1 : 0;
+  return transaction;
 }
 export function toTransactionArray(serializedTransaction) {
-    return fromRlp(`0x${serializedTransaction.slice(4)}`, 'hex');
+  return fromRlp(`0x${serializedTransaction.slice(4)}`, 'hex');
 }
 export function parseAccessList(accessList_) {
-    const accessList = [];
-    for (let i = 0; i < accessList_.length; i++) {
-        const [address, storageKeys] = accessList_[i];
-        if (!isAddress(address, { strict: false }))
-            throw new InvalidAddressError({ address });
-        accessList.push({
-            address: address,
-            storageKeys: storageKeys.map((key) => (isHash(key) ? key : trim(key))),
-        });
-    }
-    return accessList;
+  const accessList = [];
+  for (let i = 0; i < accessList_.length; i++) {
+    const [address, storageKeys] = accessList_[i];
+    if (!isAddress(address, { strict: false }))
+      throw new InvalidAddressError({ address });
+    accessList.push({
+      address: address,
+      storageKeys: storageKeys.map((key) => (isHash(key) ? key : trim(key))),
+    });
+  }
+  return accessList;
 }
 function parseAuthorizationList(serializedAuthorizationList) {
-    const authorizationList = [];
-    for (let i = 0; i < serializedAuthorizationList.length; i++) {
-        const [chainId, contractAddress, nonce, yParity, r, s] = serializedAuthorizationList[i];
-        authorizationList.push({
-            chainId: hexToNumber(chainId),
-            contractAddress,
-            nonce: hexToNumber(nonce),
-            ...parseEIP155Signature([yParity, r, s]),
-        });
-    }
-    return authorizationList;
+  const authorizationList = [];
+  for (let i = 0; i < serializedAuthorizationList.length; i++) {
+    const [chainId, contractAddress, nonce, yParity, r, s] =
+      serializedAuthorizationList[i];
+    authorizationList.push({
+      chainId: hexToNumber(chainId),
+      contractAddress,
+      nonce: hexToNumber(nonce),
+      ...parseEIP155Signature([yParity, r, s]),
+    });
+  }
+  return authorizationList;
 }
 function parseEIP155Signature(transactionArray) {
-    const signature = transactionArray.slice(-3);
-    const v = signature[0] === '0x' || hexToBigInt(signature[0]) === 0n ? 27n : 28n;
-    return {
-        r: padHex(signature[1], { size: 32 }),
-        s: padHex(signature[2], { size: 32 }),
-        v,
-        yParity: v === 27n ? 0 : 1,
-    };
+  const signature = transactionArray.slice(-3);
+  const v =
+    signature[0] === '0x' || hexToBigInt(signature[0]) === 0n ? 27n : 28n;
+  return {
+    r: padHex(signature[1], { size: 32 }),
+    s: padHex(signature[2], { size: 32 }),
+    v,
+    yParity: v === 27n ? 0 : 1,
+  };
 }
 //# sourceMappingURL=parseTransaction.js.map
diff --git a/_esm/utils/transaction/serializeTransaction.js b/_esm/utils/transaction/serializeTransaction.js
index 0a305af80eae7bdf1effb4fb3891fc6ff074fc18..7581e6667f6135a67b0119148b562b6912c9af1e 100644
--- a/_esm/utils/transaction/serializeTransaction.js
+++ b/_esm/utils/transaction/serializeTransaction.js
@@ -1,219 +1,264 @@
-import { InvalidLegacyVError, } from '../../errors/transaction.js';
-import { blobsToCommitments, } from '../blob/blobsToCommitments.js';
-import { blobsToProofs, } from '../blob/blobsToProofs.js';
-import { commitmentsToVersionedHashes, } from '../blob/commitmentsToVersionedHashes.js';
-import { toBlobSidecars, } from '../blob/toBlobSidecars.js';
+import { InvalidLegacyVError } from '../../errors/transaction.js';
+import { blobsToCommitments } from '../blob/blobsToCommitments.js';
+import { blobsToProofs } from '../blob/blobsToProofs.js';
+import { commitmentsToVersionedHashes } from '../blob/commitmentsToVersionedHashes.js';
+import { toBlobSidecars } from '../blob/toBlobSidecars.js';
 import { concatHex } from '../data/concat.js';
 import { trim } from '../data/trim.js';
 import { bytesToHex, toHex } from '../encoding/toHex.js';
 import { toRlp } from '../encoding/toRlp.js';
-import { serializeAuthorizationList, } from '../../experimental/eip7702/utils/serializeAuthorizationList.js';
-import { assertTransactionEIP1559, assertTransactionEIP2930, assertTransactionEIP4844, assertTransactionEIP7702, assertTransactionLegacy, } from './assertTransaction.js';
-import { getTransactionType, } from './getTransactionType.js';
-import { serializeAccessList, } from './serializeAccessList.js';
+import { serializeAuthorizationList } from '../../experimental/eip7702/utils/serializeAuthorizationList.js';
+import {
+  assertTransactionEIP1559,
+  assertTransactionEIP2930,
+  assertTransactionEIP4844,
+  assertTransactionEIP7702,
+  assertTransactionLegacy,
+} from './assertTransaction.js';
+import { getTransactionType } from './getTransactionType.js';
+import { serializeAccessList } from './serializeAccessList.js';
 export function serializeTransaction(transaction, signature) {
-    const type = getTransactionType(transaction);
-    if (type === 'eip1559')
-        return serializeTransactionEIP1559(transaction, signature);
-    if (type === 'eip2930')
-        return serializeTransactionEIP2930(transaction, signature);
-    if (type === 'eip4844')
-        return serializeTransactionEIP4844(transaction, signature);
-    if (type === 'eip7702')
-        return serializeTransactionEIP7702(transaction, signature);
-    return serializeTransactionLegacy(transaction, signature);
+  const type = getTransactionType(transaction);
+  if (type === 'eip1559')
+    return serializeTransactionEIP1559(transaction, signature);
+  if (type === 'eip2930')
+    return serializeTransactionEIP2930(transaction, signature);
+  if (type === 'eip4844')
+    return serializeTransactionEIP4844(transaction, signature);
+  if (type === 'eip7702')
+    return serializeTransactionEIP7702(transaction, signature);
+  return serializeTransactionLegacy(transaction, signature);
 }
 function serializeTransactionEIP7702(transaction, signature) {
-    const { authorizationList, chainId, gas, nonce, to, value, maxFeePerGas, maxPriorityFeePerGas, accessList, data, } = transaction;
-    assertTransactionEIP7702(transaction);
-    const serializedAccessList = serializeAccessList(accessList);
-    const serializedAuthorizationList = serializeAuthorizationList(authorizationList);
-    return concatHex([
-        '0x04',
-        toRlp([
-            toHex(chainId),
-            nonce ? toHex(nonce) : '0x',
-            maxPriorityFeePerGas ? toHex(maxPriorityFeePerGas) : '0x',
-            maxFeePerGas ? toHex(maxFeePerGas) : '0x',
-            gas ? toHex(gas) : '0x',
-            to ?? '0x',
-            value ? toHex(value) : '0x',
-            data ?? '0x',
-            serializedAccessList,
-            serializedAuthorizationList,
-            ...toYParitySignatureArray(transaction, signature),
-        ]),
-    ]);
+  const {
+    authorizationList,
+    chainId,
+    gas,
+    nonce,
+    to,
+    value,
+    maxFeePerGas,
+    maxPriorityFeePerGas,
+    accessList,
+    data,
+  } = transaction;
+  assertTransactionEIP7702(transaction);
+  const serializedAccessList = serializeAccessList(accessList);
+  const serializedAuthorizationList =
+    serializeAuthorizationList(authorizationList);
+  return concatHex([
+    '0x04',
+    toRlp([
+      toHex(chainId),
+      nonce ? toHex(nonce) : '0x',
+      maxPriorityFeePerGas ? toHex(maxPriorityFeePerGas) : '0x',
+      maxFeePerGas ? toHex(maxFeePerGas) : '0x',
+      gas ? toHex(gas) : '0x',
+      to ?? '0x',
+      value ? toHex(value) : '0x',
+      data ?? '0x',
+      serializedAccessList,
+      serializedAuthorizationList,
+      ...toYParitySignatureArray(transaction, signature),
+    ]),
+  ]);
 }
 function serializeTransactionEIP4844(transaction, signature) {
-    const { chainId, gas, nonce, to, value, maxFeePerBlobGas, maxFeePerGas, maxPriorityFeePerGas, accessList, data, } = transaction;
-    assertTransactionEIP4844(transaction);
-    let blobVersionedHashes = transaction.blobVersionedHashes;
-    let sidecars = transaction.sidecars;
-    // If `blobs` are passed, we will need to compute the KZG commitments & proofs.
-    if (transaction.blobs &&
-        (typeof blobVersionedHashes === 'undefined' ||
-            typeof sidecars === 'undefined')) {
-        const blobs = (typeof transaction.blobs[0] === 'string'
-            ? transaction.blobs
-            : transaction.blobs.map((x) => bytesToHex(x)));
-        const kzg = transaction.kzg;
-        const commitments = blobsToCommitments({
-            blobs,
-            kzg,
-        });
-        if (typeof blobVersionedHashes === 'undefined')
-            blobVersionedHashes = commitmentsToVersionedHashes({
-                commitments,
-            });
-        if (typeof sidecars === 'undefined') {
-            const proofs = blobsToProofs({ blobs, commitments, kzg });
-            sidecars = toBlobSidecars({ blobs, commitments, proofs });
-        }
+  const {
+    chainId,
+    gas,
+    nonce,
+    to,
+    value,
+    maxFeePerBlobGas,
+    maxFeePerGas,
+    maxPriorityFeePerGas,
+    accessList,
+    data,
+  } = transaction;
+  assertTransactionEIP4844(transaction);
+  let blobVersionedHashes = transaction.blobVersionedHashes;
+  let sidecars = transaction.sidecars;
+  // If `blobs` are passed, we will need to compute the KZG commitments & proofs.
+  if (
+    transaction.blobs &&
+    (typeof blobVersionedHashes === 'undefined' ||
+      typeof sidecars === 'undefined')
+  ) {
+    const blobs =
+      typeof transaction.blobs[0] === 'string'
+        ? transaction.blobs
+        : transaction.blobs.map((x) => bytesToHex(x));
+    const kzg = transaction.kzg;
+    const commitments = blobsToCommitments({
+      blobs,
+      kzg,
+    });
+    if (typeof blobVersionedHashes === 'undefined')
+      blobVersionedHashes = commitmentsToVersionedHashes({
+        commitments,
+      });
+    if (typeof sidecars === 'undefined') {
+      const proofs = blobsToProofs({ blobs, commitments, kzg });
+      sidecars = toBlobSidecars({ blobs, commitments, proofs });
     }
-    const serializedAccessList = serializeAccessList(accessList);
-    const serializedTransaction = [
-        toHex(chainId),
-        nonce ? toHex(nonce) : '0x',
-        maxPriorityFeePerGas ? toHex(maxPriorityFeePerGas) : '0x',
-        maxFeePerGas ? toHex(maxFeePerGas) : '0x',
-        gas ? toHex(gas) : '0x',
-        to ?? '0x',
-        value ? toHex(value) : '0x',
-        data ?? '0x',
-        serializedAccessList,
-        maxFeePerBlobGas ? toHex(maxFeePerBlobGas) : '0x',
-        blobVersionedHashes ?? [],
-        ...toYParitySignatureArray(transaction, signature),
-    ];
-    const blobs = [];
-    const commitments = [];
-    const proofs = [];
-    if (sidecars)
-        for (let i = 0; i < sidecars.length; i++) {
-            const { blob, commitment, proof } = sidecars[i];
-            blobs.push(blob);
-            commitments.push(commitment);
-            proofs.push(proof);
-        }
+  }
+  const serializedAccessList = serializeAccessList(accessList);
+  const serializedTransaction = [
+    toHex(chainId),
+    nonce ? toHex(nonce) : '0x',
+    maxPriorityFeePerGas ? toHex(maxPriorityFeePerGas) : '0x',
+    maxFeePerGas ? toHex(maxFeePerGas) : '0x',
+    gas ? toHex(gas) : '0x',
+    to ?? '0x',
+    value ? toHex(value) : '0x',
+    data ?? '0x',
+    serializedAccessList,
+    maxFeePerBlobGas ? toHex(maxFeePerBlobGas) : '0x',
+    blobVersionedHashes ?? [],
+    ...toYParitySignatureArray(transaction, signature),
+  ];
+  const blobs = [];
+  const commitments = [];
+  const proofs = [];
+  if (sidecars)
+    for (let i = 0; i < sidecars.length; i++) {
+      const { blob, commitment, proof } = sidecars[i];
+      blobs.push(blob);
+      commitments.push(commitment);
+      // proof can be a single proof (EIP-4844) or an array of proofs (EIP-7594)
+      if (Array.isArray(proof)) {
+        proofs.push(...proof);
+      } else {
+        proofs.push(proof);
+      }
+    }
+
+  if (chainId === 11155111) {
     return concatHex([
-        '0x03',
-        sidecars
-            ? // If sidecars are enabled, envelope turns into a "wrapper":
-                toRlp([serializedTransaction, blobs, commitments, proofs])
-            : // If sidecars are disabled, standard envelope is used:
-                toRlp(serializedTransaction),
+      '0x03',
+      sidecars
+        ? // If sidecars are enabled, envelope turns into a "wrapper":
+          toRlp([serializedTransaction, '0x01', blobs, commitments, proofs])
+        : // If sidecars are disabled, standard envelope is used:
+          toRlp(serializedTransaction),
     ]);
+  }
+
+  return concatHex([
+    '0x03',
+    sidecars
+      ? // If sidecars are enabled, envelope turns into a "wrapper":
+        toRlp([serializedTransaction, blobs, commitments, proofs])
+      : // If sidecars are disabled, standard envelope is used:
+        toRlp(serializedTransaction),
+  ]);
 }
 function serializeTransactionEIP1559(transaction, signature) {
-    const { chainId, gas, nonce, to, value, maxFeePerGas, maxPriorityFeePerGas, accessList, data, } = transaction;
-    assertTransactionEIP1559(transaction);
-    const serializedAccessList = serializeAccessList(accessList);
-    const serializedTransaction = [
-        toHex(chainId),
-        nonce ? toHex(nonce) : '0x',
-        maxPriorityFeePerGas ? toHex(maxPriorityFeePerGas) : '0x',
-        maxFeePerGas ? toHex(maxFeePerGas) : '0x',
-        gas ? toHex(gas) : '0x',
-        to ?? '0x',
-        value ? toHex(value) : '0x',
-        data ?? '0x',
-        serializedAccessList,
-        ...toYParitySignatureArray(transaction, signature),
-    ];
-    return concatHex([
-        '0x02',
-        toRlp(serializedTransaction),
-    ]);
+  const {
+    chainId,
+    gas,
+    nonce,
+    to,
+    value,
+    maxFeePerGas,
+    maxPriorityFeePerGas,
+    accessList,
+    data,
+  } = transaction;
+  assertTransactionEIP1559(transaction);
+  const serializedAccessList = serializeAccessList(accessList);
+  const serializedTransaction = [
+    toHex(chainId),
+    nonce ? toHex(nonce) : '0x',
+    maxPriorityFeePerGas ? toHex(maxPriorityFeePerGas) : '0x',
+    maxFeePerGas ? toHex(maxFeePerGas) : '0x',
+    gas ? toHex(gas) : '0x',
+    to ?? '0x',
+    value ? toHex(value) : '0x',
+    data ?? '0x',
+    serializedAccessList,
+    ...toYParitySignatureArray(transaction, signature),
+  ];
+  return concatHex(['0x02', toRlp(serializedTransaction)]);
 }
 function serializeTransactionEIP2930(transaction, signature) {
-    const { chainId, gas, data, nonce, to, value, accessList, gasPrice } = transaction;
-    assertTransactionEIP2930(transaction);
-    const serializedAccessList = serializeAccessList(accessList);
-    const serializedTransaction = [
-        toHex(chainId),
-        nonce ? toHex(nonce) : '0x',
-        gasPrice ? toHex(gasPrice) : '0x',
-        gas ? toHex(gas) : '0x',
-        to ?? '0x',
-        value ? toHex(value) : '0x',
-        data ?? '0x',
-        serializedAccessList,
-        ...toYParitySignatureArray(transaction, signature),
-    ];
-    return concatHex([
-        '0x01',
-        toRlp(serializedTransaction),
-    ]);
+  const { chainId, gas, data, nonce, to, value, accessList, gasPrice } =
+    transaction;
+  assertTransactionEIP2930(transaction);
+  const serializedAccessList = serializeAccessList(accessList);
+  const serializedTransaction = [
+    toHex(chainId),
+    nonce ? toHex(nonce) : '0x',
+    gasPrice ? toHex(gasPrice) : '0x',
+    gas ? toHex(gas) : '0x',
+    to ?? '0x',
+    value ? toHex(value) : '0x',
+    data ?? '0x',
+    serializedAccessList,
+    ...toYParitySignatureArray(transaction, signature),
+  ];
+  return concatHex(['0x01', toRlp(serializedTransaction)]);
 }
 function serializeTransactionLegacy(transaction, signature) {
-    const { chainId = 0, gas, data, nonce, to, value, gasPrice } = transaction;
-    assertTransactionLegacy(transaction);
-    let serializedTransaction = [
-        nonce ? toHex(nonce) : '0x',
-        gasPrice ? toHex(gasPrice) : '0x',
-        gas ? toHex(gas) : '0x',
-        to ?? '0x',
-        value ? toHex(value) : '0x',
-        data ?? '0x',
+  const { chainId = 0, gas, data, nonce, to, value, gasPrice } = transaction;
+  assertTransactionLegacy(transaction);
+  let serializedTransaction = [
+    nonce ? toHex(nonce) : '0x',
+    gasPrice ? toHex(gasPrice) : '0x',
+    gas ? toHex(gas) : '0x',
+    to ?? '0x',
+    value ? toHex(value) : '0x',
+    data ?? '0x',
+  ];
+  if (signature) {
+    const v = (() => {
+      // EIP-155 (inferred chainId)
+      if (signature.v >= 35n) {
+        const inferredChainId = (signature.v - 35n) / 2n;
+        if (inferredChainId > 0) return signature.v;
+        return 27n + (signature.v === 35n ? 0n : 1n);
+      }
+      // EIP-155 (explicit chainId)
+      if (chainId > 0)
+        return BigInt(chainId * 2) + BigInt(35n + signature.v - 27n);
+      // Pre-EIP-155 (no chainId)
+      const v = 27n + (signature.v === 27n ? 0n : 1n);
+      if (signature.v !== v) throw new InvalidLegacyVError({ v: signature.v });
+      return v;
+    })();
+    const r = trim(signature.r);
+    const s = trim(signature.s);
+    serializedTransaction = [
+      ...serializedTransaction,
+      toHex(v),
+      r === '0x00' ? '0x' : r,
+      s === '0x00' ? '0x' : s,
     ];
-    if (signature) {
-        const v = (() => {
-            // EIP-155 (inferred chainId)
-            if (signature.v >= 35n) {
-                const inferredChainId = (signature.v - 35n) / 2n;
-                if (inferredChainId > 0)
-                    return signature.v;
-                return 27n + (signature.v === 35n ? 0n : 1n);
-            }
-            // EIP-155 (explicit chainId)
-            if (chainId > 0)
-                return BigInt(chainId * 2) + BigInt(35n + signature.v - 27n);
-            // Pre-EIP-155 (no chainId)
-            const v = 27n + (signature.v === 27n ? 0n : 1n);
-            if (signature.v !== v)
-                throw new InvalidLegacyVError({ v: signature.v });
-            return v;
-        })();
-        const r = trim(signature.r);
-        const s = trim(signature.s);
-        serializedTransaction = [
-            ...serializedTransaction,
-            toHex(v),
-            r === '0x00' ? '0x' : r,
-            s === '0x00' ? '0x' : s,
-        ];
-    }
-    else if (chainId > 0) {
-        serializedTransaction = [
-            ...serializedTransaction,
-            toHex(chainId),
-            '0x',
-            '0x',
-        ];
-    }
-    return toRlp(serializedTransaction);
+  } else if (chainId > 0) {
+    serializedTransaction = [
+      ...serializedTransaction,
+      toHex(chainId),
+      '0x',
+      '0x',
+    ];
+  }
+  return toRlp(serializedTransaction);
 }
 export function toYParitySignatureArray(transaction, signature_) {
-    const signature = signature_ ?? transaction;
-    const { v, yParity } = signature;
-    if (typeof signature.r === 'undefined')
-        return [];
-    if (typeof signature.s === 'undefined')
-        return [];
-    if (typeof v === 'undefined' && typeof yParity === 'undefined')
-        return [];
-    const r = trim(signature.r);
-    const s = trim(signature.s);
-    const yParity_ = (() => {
-        if (typeof yParity === 'number')
-            return yParity ? toHex(1) : '0x';
-        if (v === 0n)
-            return '0x';
-        if (v === 1n)
-            return toHex(1);
-        return v === 27n ? '0x' : toHex(1);
-    })();
-    return [yParity_, r === '0x00' ? '0x' : r, s === '0x00' ? '0x' : s];
+  const signature = signature_ ?? transaction;
+  const { v, yParity } = signature;
+  if (typeof signature.r === 'undefined') return [];
+  if (typeof signature.s === 'undefined') return [];
+  if (typeof v === 'undefined' && typeof yParity === 'undefined') return [];
+  const r = trim(signature.r);
+  const s = trim(signature.s);
+  const yParity_ = (() => {
+    if (typeof yParity === 'number') return yParity ? toHex(1) : '0x';
+    if (v === 0n) return '0x';
+    if (v === 1n) return toHex(1);
+    return v === 27n ? '0x' : toHex(1);
+  })();
+  return [yParity_, r === '0x00' ? '0x' : r, s === '0x00' ? '0x' : s];
 }
 //# sourceMappingURL=serializeTransaction.js.map
diff --git a/package.json b/package.json
index c119f5955094a66a4807f94e496a0f9009de2102..7003b57a45c09a067f1bc16bc85263d1154ad127 100644
--- a/package.json
+++ b/package.json
@@ -203,5 +203,6 @@
     "wallet",
     "web3",
     "typescript"
-  ]
+  ],
+  "packageManager": "yarn@4.10.3+sha512.c38cafb5c7bb273f3926d04e55e1d8c9dfa7d9c3ea1f36a4868fa028b9e5f72298f0b7f401ad5eb921749eb012eb1c3bb74bf7503df3ee43fd600d14a018266f"
 }
diff --git a/utils/transaction/serializeTransaction.ts b/utils/transaction/serializeTransaction.ts
index a8aa0baf074fb39f5ddfad0b683e71740464e155..a60a98f6894c347bf1319ae0c41d00e798a0ed33 100644
--- a/utils/transaction/serializeTransaction.ts
+++ b/utils/transaction/serializeTransaction.ts
@@ -1,14 +1,14 @@
 import {
   InvalidLegacyVError,
   type InvalidLegacyVErrorType,
-} from '../../errors/transaction.js'
-import type { ErrorType } from '../../errors/utils.js'
+} from '../../errors/transaction.js';
+import type { ErrorType } from '../../errors/utils.js';
 import type {
   ByteArray,
   Hex,
   Signature,
   SignatureLegacy,
-} from '../../types/misc.js'
+} from '../../types/misc.js';
 import type {
   TransactionSerializable,
   TransactionSerializableEIP1559,
@@ -24,33 +24,33 @@ import type {
   TransactionSerializedEIP7702,
   TransactionSerializedLegacy,
   TransactionType,
-} from '../../types/transaction.js'
-import type { OneOf } from '../../types/utils.js'
+} from '../../types/transaction.js';
+import type { OneOf } from '../../types/utils.js';
 import {
   type BlobsToCommitmentsErrorType,
   blobsToCommitments,
-} from '../blob/blobsToCommitments.js'
+} from '../blob/blobsToCommitments.js';
 import {
   blobsToProofs,
   type blobsToProofsErrorType,
-} from '../blob/blobsToProofs.js'
+} from '../blob/blobsToProofs.js';
 import {
   type CommitmentsToVersionedHashesErrorType,
   commitmentsToVersionedHashes,
-} from '../blob/commitmentsToVersionedHashes.js'
+} from '../blob/commitmentsToVersionedHashes.js';
 import {
   type ToBlobSidecarsErrorType,
   toBlobSidecars,
-} from '../blob/toBlobSidecars.js'
-import { type ConcatHexErrorType, concatHex } from '../data/concat.js'
-import { trim } from '../data/trim.js'
-import { type ToHexErrorType, bytesToHex, toHex } from '../encoding/toHex.js'
-import { type ToRlpErrorType, toRlp } from '../encoding/toRlp.js'
+} from '../blob/toBlobSidecars.js';
+import { type ConcatHexErrorType, concatHex } from '../data/concat.js';
+import { trim } from '../data/trim.js';
+import { type ToHexErrorType, bytesToHex, toHex } from '../encoding/toHex.js';
+import { type ToRlpErrorType, toRlp } from '../encoding/toRlp.js';
 
 import {
   type SerializeAuthorizationListErrorType,
   serializeAuthorizationList,
-} from '../../experimental/eip7702/utils/serializeAuthorizationList.js'
+} from '../../experimental/eip7702/utils/serializeAuthorizationList.js';
 import {
   type AssertTransactionEIP1559ErrorType,
   type AssertTransactionEIP2930ErrorType,
@@ -62,22 +62,22 @@ import {
   assertTransactionEIP4844,
   assertTransactionEIP7702,
   assertTransactionLegacy,
-} from './assertTransaction.js'
+} from './assertTransaction.js';
 import {
   type GetTransactionType,
   type GetTransactionTypeErrorType,
   getTransactionType,
-} from './getTransactionType.js'
+} from './getTransactionType.js';
 import {
   type SerializeAccessListErrorType,
   serializeAccessList,
-} from './serializeAccessList.js'
+} from './serializeAccessList.js';
 
 export type SerializedTransactionReturnType<
   transaction extends TransactionSerializable = TransactionSerializable,
   ///
   _transactionType extends TransactionType = GetTransactionType<transaction>,
-> = TransactionSerialized<_transactionType>
+> = TransactionSerialized<_transactionType>;
 
 export type SerializeTransactionFn<
   transaction extends TransactionSerializableGeneric = TransactionSerializable,
@@ -86,7 +86,7 @@ export type SerializeTransactionFn<
 > = typeof serializeTransaction<
   OneOf<TransactionSerializable | transaction>,
   _transactionType
->
+>;
 
 export type SerializeTransactionErrorType =
   | GetTransactionTypeErrorType
@@ -95,7 +95,7 @@ export type SerializeTransactionErrorType =
   | SerializeTransactionEIP4844ErrorType
   | SerializeTransactionEIP7702ErrorType
   | SerializeTransactionLegacyErrorType
-  | ErrorType
+  | ErrorType;
 
 export function serializeTransaction<
   const transaction extends TransactionSerializable,
@@ -105,36 +105,36 @@ export function serializeTransaction<
   transaction: transaction,
   signature?: Signature | undefined,
 ): SerializedTransactionReturnType<transaction, _transactionType> {
-  const type = getTransactionType(transaction) as GetTransactionType
+  const type = getTransactionType(transaction) as GetTransactionType;
 
   if (type === 'eip1559')
     return serializeTransactionEIP1559(
       transaction as TransactionSerializableEIP1559,
       signature,
-    ) as SerializedTransactionReturnType<transaction>
+    ) as SerializedTransactionReturnType<transaction>;
 
   if (type === 'eip2930')
     return serializeTransactionEIP2930(
       transaction as TransactionSerializableEIP2930,
       signature,
-    ) as SerializedTransactionReturnType<transaction>
+    ) as SerializedTransactionReturnType<transaction>;
 
   if (type === 'eip4844')
     return serializeTransactionEIP4844(
       transaction as TransactionSerializableEIP4844,
       signature,
-    ) as SerializedTransactionReturnType<transaction>
+    ) as SerializedTransactionReturnType<transaction>;
 
   if (type === 'eip7702')
     return serializeTransactionEIP7702(
       transaction as TransactionSerializableEIP7702,
       signature,
-    ) as SerializedTransactionReturnType<transaction>
+    ) as SerializedTransactionReturnType<transaction>;
 
   return serializeTransactionLegacy(
     transaction as TransactionSerializableLegacy,
     signature as SignatureLegacy,
-  ) as SerializedTransactionReturnType<transaction>
+  ) as SerializedTransactionReturnType<transaction>;
 }
 
 type SerializeTransactionEIP7702ErrorType =
@@ -145,7 +145,7 @@ type SerializeTransactionEIP7702ErrorType =
   | ToHexErrorType
   | ToRlpErrorType
   | SerializeAccessListErrorType
-  | ErrorType
+  | ErrorType;
 
 function serializeTransactionEIP7702(
   transaction: TransactionSerializableEIP7702,
@@ -162,13 +162,13 @@ function serializeTransactionEIP7702(
     maxPriorityFeePerGas,
     accessList,
     data,
-  } = transaction
+  } = transaction;
 
-  assertTransactionEIP7702(transaction)
+  assertTransactionEIP7702(transaction);
 
-  const serializedAccessList = serializeAccessList(accessList)
+  const serializedAccessList = serializeAccessList(accessList);
   const serializedAuthorizationList =
-    serializeAuthorizationList(authorizationList)
+    serializeAuthorizationList(authorizationList);
 
   return concatHex([
     '0x04',
@@ -185,7 +185,7 @@ function serializeTransactionEIP7702(
       serializedAuthorizationList,
       ...toYParitySignatureArray(transaction, signature),
     ]),
-  ]) as TransactionSerializedEIP7702
+  ]) as TransactionSerializedEIP7702;
 }
 
 type SerializeTransactionEIP4844ErrorType =
@@ -199,7 +199,7 @@ type SerializeTransactionEIP4844ErrorType =
   | ToHexErrorType
   | ToRlpErrorType
   | SerializeAccessListErrorType
-  | ErrorType
+  | ErrorType;
 
 function serializeTransactionEIP4844(
   transaction: TransactionSerializableEIP4844,
@@ -216,12 +216,12 @@ function serializeTransactionEIP4844(
     maxPriorityFeePerGas,
     accessList,
     data,
-  } = transaction
+  } = transaction;
 
-  assertTransactionEIP4844(transaction)
+  assertTransactionEIP4844(transaction);
 
-  let blobVersionedHashes = transaction.blobVersionedHashes
-  let sidecars = transaction.sidecars
+  let blobVersionedHashes = transaction.blobVersionedHashes;
+  let sidecars = transaction.sidecars;
   // If `blobs` are passed, we will need to compute the KZG commitments & proofs.
   if (
     transaction.blobs &&
@@ -232,24 +232,24 @@ function serializeTransactionEIP4844(
       typeof transaction.blobs[0] === 'string'
         ? transaction.blobs
         : (transaction.blobs as ByteArray[]).map((x) => bytesToHex(x))
-    ) as Hex[]
-    const kzg = transaction.kzg!
+    ) as Hex[];
+    const kzg = transaction.kzg!;
     const commitments = blobsToCommitments({
       blobs,
       kzg,
-    })
+    });
 
     if (typeof blobVersionedHashes === 'undefined')
       blobVersionedHashes = commitmentsToVersionedHashes({
         commitments,
-      })
+      });
     if (typeof sidecars === 'undefined') {
-      const proofs = blobsToProofs({ blobs, commitments, kzg })
-      sidecars = toBlobSidecars({ blobs, commitments, proofs })
+      const proofs = blobsToProofs({ blobs, commitments, kzg });
+      sidecars = toBlobSidecars({ blobs, commitments, proofs });
     }
   }
 
-  const serializedAccessList = serializeAccessList(accessList)
+  const serializedAccessList = serializeAccessList(accessList);
 
   const serializedTransaction = [
     toHex(chainId),
@@ -264,27 +264,33 @@ function serializeTransactionEIP4844(
     maxFeePerBlobGas ? toHex(maxFeePerBlobGas) : '0x',
     blobVersionedHashes ?? [],
     ...toYParitySignatureArray(transaction, signature),
-  ] as const
+  ] as const;
 
-  const blobs: Hex[] = []
-  const commitments: Hex[] = []
-  const proofs: Hex[] = []
+  const blobs: Hex[] = [];
+  const commitments: Hex[] = [];
+  const proofs: Hex[] = [];
   if (sidecars)
     for (let i = 0; i < sidecars.length; i++) {
-      const { blob, commitment, proof } = sidecars[i]
-      blobs.push(blob)
-      commitments.push(commitment)
-      proofs.push(proof)
+      const { blob, commitment, proof } = sidecars[i];
+      blobs.push(blob);
+      commitments.push(commitment);
+      proofs.push(...(proof as unknown as Hex[]));
     }
 
   return concatHex([
     '0x03',
     sidecars
       ? // If sidecars are enabled, envelope turns into a "wrapper":
-        toRlp([serializedTransaction, blobs, commitments, proofs])
+        toRlp([
+          serializedTransaction,
+          '0x01',
+          blobs,
+          commitments,
+          `0x${proofs.map((x) => x.replace('0x', '')).join('')}`,
+        ])
       : // If sidecars are disabled, standard envelope is used:
         toRlp(serializedTransaction),
-  ]) as TransactionSerializedEIP4844
+  ]) as TransactionSerializedEIP4844;
 }
 
 type SerializeTransactionEIP1559ErrorType =
@@ -294,7 +300,7 @@ type SerializeTransactionEIP1559ErrorType =
   | ToHexErrorType
   | ToRlpErrorType
   | SerializeAccessListErrorType
-  | ErrorType
+  | ErrorType;
 
 function serializeTransactionEIP1559(
   transaction: TransactionSerializableEIP1559,
@@ -310,11 +316,11 @@ function serializeTransactionEIP1559(
     maxPriorityFeePerGas,
     accessList,
     data,
-  } = transaction
+  } = transaction;
 
-  assertTransactionEIP1559(transaction)
+  assertTransactionEIP1559(transaction);
 
-  const serializedAccessList = serializeAccessList(accessList)
+  const serializedAccessList = serializeAccessList(accessList);
 
   const serializedTransaction = [
     toHex(chainId),
@@ -327,12 +333,12 @@ function serializeTransactionEIP1559(
     data ?? '0x',
     serializedAccessList,
     ...toYParitySignatureArray(transaction, signature),
-  ]
+  ];
 
   return concatHex([
     '0x02',
     toRlp(serializedTransaction),
-  ]) as TransactionSerializedEIP1559
+  ]) as TransactionSerializedEIP1559;
 }
 
 type SerializeTransactionEIP2930ErrorType =
@@ -342,18 +348,18 @@ type SerializeTransactionEIP2930ErrorType =
   | ToHexErrorType
   | ToRlpErrorType
   | SerializeAccessListErrorType
-  | ErrorType
+  | ErrorType;
 
 function serializeTransactionEIP2930(
   transaction: TransactionSerializableEIP2930,
   signature?: Signature | undefined,
 ): TransactionSerializedEIP2930 {
   const { chainId, gas, data, nonce, to, value, accessList, gasPrice } =
-    transaction
+    transaction;
 
-  assertTransactionEIP2930(transaction)
+  assertTransactionEIP2930(transaction);
 
-  const serializedAccessList = serializeAccessList(accessList)
+  const serializedAccessList = serializeAccessList(accessList);
 
   const serializedTransaction = [
     toHex(chainId),
@@ -365,12 +371,12 @@ function serializeTransactionEIP2930(
     data ?? '0x',
     serializedAccessList,
     ...toYParitySignatureArray(transaction, signature),
-  ]
+  ];
 
   return concatHex([
     '0x01',
     toRlp(serializedTransaction),
-  ]) as TransactionSerializedEIP2930
+  ]) as TransactionSerializedEIP2930;
 }
 
 type SerializeTransactionLegacyErrorType =
@@ -378,15 +384,15 @@ type SerializeTransactionLegacyErrorType =
   | InvalidLegacyVErrorType
   | ToHexErrorType
   | ToRlpErrorType
-  | ErrorType
+  | ErrorType;
 
 function serializeTransactionLegacy(
   transaction: TransactionSerializableLegacy,
   signature?: SignatureLegacy | undefined,
 ): TransactionSerializedLegacy {
-  const { chainId = 0, gas, data, nonce, to, value, gasPrice } = transaction
+  const { chainId = 0, gas, data, nonce, to, value, gasPrice } = transaction;
 
-  assertTransactionLegacy(transaction)
+  assertTransactionLegacy(transaction);
 
   let serializedTransaction = [
     nonce ? toHex(nonce) : '0x',
@@ -395,69 +401,69 @@ function serializeTransactionLegacy(
     to ?? '0x',
     value ? toHex(value) : '0x',
     data ?? '0x',
-  ]
+  ];
 
   if (signature) {
     const v = (() => {
       // EIP-155 (inferred chainId)
       if (signature.v >= 35n) {
-        const inferredChainId = (signature.v - 35n) / 2n
-        if (inferredChainId > 0) return signature.v
-        return 27n + (signature.v === 35n ? 0n : 1n)
+        const inferredChainId = (signature.v - 35n) / 2n;
+        if (inferredChainId > 0) return signature.v;
+        return 27n + (signature.v === 35n ? 0n : 1n);
       }
 
       // EIP-155 (explicit chainId)
       if (chainId > 0)
-        return BigInt(chainId * 2) + BigInt(35n + signature.v - 27n)
+        return BigInt(chainId * 2) + BigInt(35n + signature.v - 27n);
 
       // Pre-EIP-155 (no chainId)
-      const v = 27n + (signature.v === 27n ? 0n : 1n)
-      if (signature.v !== v) throw new InvalidLegacyVError({ v: signature.v })
-      return v
-    })()
+      const v = 27n + (signature.v === 27n ? 0n : 1n);
+      if (signature.v !== v) throw new InvalidLegacyVError({ v: signature.v });
+      return v;
+    })();
 
-    const r = trim(signature.r)
-    const s = trim(signature.s)
+    const r = trim(signature.r);
+    const s = trim(signature.s);
 
     serializedTransaction = [
       ...serializedTransaction,
       toHex(v),
       r === '0x00' ? '0x' : r,
       s === '0x00' ? '0x' : s,
-    ]
+    ];
   } else if (chainId > 0) {
     serializedTransaction = [
       ...serializedTransaction,
       toHex(chainId),
       '0x',
       '0x',
-    ]
+    ];
   }
 
-  return toRlp(serializedTransaction) as TransactionSerializedLegacy
+  return toRlp(serializedTransaction) as TransactionSerializedLegacy;
 }
 
 export function toYParitySignatureArray(
   transaction: TransactionSerializableGeneric,
   signature_?: Signature | undefined,
 ) {
-  const signature = signature_ ?? transaction
-  const { v, yParity } = signature
+  const signature = signature_ ?? transaction;
+  const { v, yParity } = signature;
 
-  if (typeof signature.r === 'undefined') return []
-  if (typeof signature.s === 'undefined') return []
-  if (typeof v === 'undefined' && typeof yParity === 'undefined') return []
+  if (typeof signature.r === 'undefined') return [];
+  if (typeof signature.s === 'undefined') return [];
+  if (typeof v === 'undefined' && typeof yParity === 'undefined') return [];
 
-  const r = trim(signature.r)
-  const s = trim(signature.s)
+  const r = trim(signature.r);
+  const s = trim(signature.s);
 
   const yParity_ = (() => {
-    if (typeof yParity === 'number') return yParity ? toHex(1) : '0x'
-    if (v === 0n) return '0x'
-    if (v === 1n) return toHex(1)
+    if (typeof yParity === 'number') return yParity ? toHex(1) : '0x';
+    if (v === 0n) return '0x';
+    if (v === 1n) return toHex(1);
 
-    return v === 27n ? '0x' : toHex(1)
-  })()
+    return v === 27n ? '0x' : toHex(1);
+  })();
 
-  return [yParity_, r === '0x00' ? '0x' : r, s === '0x00' ? '0x' : s]
+  return [yParity_, r === '0x00' ? '0x' : r, s === '0x00' ? '0x' : s];
 }
